### PR TITLE
Ground contact force visualization for SmoothSphereHalfSpaceForce

### PR DESCRIPTION
**Pull request by nickbianco**
### Brief summary of changes
Additions to SmoothSphereHalfSpaceForce to visualize contact forces (as narrow cylinders) in the Simbody visualizer.

### CHANGELOG.md (choose one)

- [x] updated
- [ ] no need to update because...

---
This change is [Reviewable](https://reviewable.io/reviews/opensim-org/opensim-moco/518)

### CODE REVIEW

File: Moco/Moco/Components/SmoothSphereHalfSpaceForce.cpp [hunk 1]
```diff
@@ -213,7 +216,66 @@ void SmoothSphereHalfSpaceForce::generateDecorations(bool fixed,
...
+
+        // Compute the force visualization scale factor.
+        double forceVizScaleFactor;
+        if (!getProperty_force_visualization_scale_factor().empty()) {
+            forceVizScaleFactor = get_force_visualization_scale_factor();
+        } else {
+            const double mass = model.getTotalMass(state);
+            const double weight = mass * model.getGravity().norm();
+            forceVizScaleFactor = 1 / weight;
+        }
```
[the remark this answers is not in the document]
Comment (nickbianco) [reply Moco/Moco/Components/SmoothSphereHalfSpaceForce.cpp]: Ah, thank you. I tried to cache in `extendAddToSystem`, but that doesn't provide a state.

File: Moco/Moco/Components/SmoothSphereHalfSpaceForce.cpp [hunk 2]
```diff
@@ -213,7 +216,66 @@ void SmoothSphereHalfSpaceForce::generateDecorations(bool fixed,
...
+
+        // Scale the contact force vector and compute the cylinder length.
+        const auto& scaledContactForce =
+                contactSphereForce * forceVizScaleFactor;
+        const SimTK::Real& length(scaledContactForce.norm());
```
[the remark this answers is not in the document]
Comment (nickbianco) [reply Moco/Moco/Components/SmoothSphereHalfSpaceForce.cpp]: Hmm, not sure how that happened.

**Comment by nickbianco** [comment]
Ready for review.

File: Moco/Moco/Components/SmoothSphereHalfSpaceForce.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (chrisdembia) [reply Moco/Moco/Components/SmoothSphereHalfSpaceForce.cpp]: Based on https://simbody.github.io/simbody-3.6-doxygen/api/classSimTK_1_1Stage.html, maybe `extendRealizeInstance()` makes more sense. Sorry for not looking into this more initially.

**Comment by chrisdembia** [comment]
One more comment. After that, LGTM.

**Comment by nickbianco** [comment]
@chrisdembia so I tried `extendRealizeInstance`, but got a compiler error:
```
error C3668: 'OpenSim::SmoothSphereHalfSpaceForce::extendRealizeInstance': method with override specifier 'override' did not override any base class methods.
```
 I'm not sure why, as `extendRealizeTopology` works fine and both are virtual methods of `Component`. I switched back to `extendRealizeTopology`; let me know if that's fine. I could submit a bug to opensim-core if that makes sense.

**Comment by chrisdembia** [comment]
@nickbianco the functions have different arguments; `Instance` takes a const state.

**Comment by nickbianco** [comment]
🤦‍♂ fixed.

**Pull request merged**